crates/pathfinder/src/state/contract_hash.rs

```diff
@@ -0,0 +1,419 @@
+                .flat_map(|(i, x)| {
+                    [("selector", &*x.selector), ("offset", &*x.offset)]
+                        .into_iter()
+                        .map(move |(field, x)| match x.strip_prefix("0x") {
+                            Some(x) => Ok((field, x)),
+                            None => Err(anyhow::anyhow!(
+                                "Entry point missing '0x' prefix under {key} at index {i} entry ({field})",
+                            )),
+                        })
+                        .map(move |res| {
+                            res.and_then(|(field, hex)| {
+                                StarkHash::from_hex_str(hex).with_context(|| {
+                                    format!("Entry point invalid hex under {key} at index {i} entry ({field})")
+                                })
+                            })
+                        })
+                })
+                .try_fold(HashChain::default(), |mut hc, next| {
+                    hc.update(next?);
+                    Result::<_, Error>::Ok(hc)
+                })
+        })
+        .try_for_each(|x| {
+            outer.update(x?.finalize());
+            Result::<_, Error>::Ok(())
+        })
+        .context("Failed to process contract_definition.entry_points_by_type")?;
+
+    let builtins = contract_definition
+        .program
+        .builtins
+        .iter()
+        .enumerate()
+        .map(|(i, s)| (i, s.as_bytes()))
+        .map(|(i, s)| {
+            StarkHash::from_be_slice(s).with_context(|| format!("Invalid builtin at index {i}"))
+        })
+        .try_fold(HashChain::default(), |mut hc, next| {
+            hc.update(next?);
+            Result::<_, Error>::Ok(hc)
+        })
+        .context("Failed to process contract_definition.program.builtins")?;
+
+    outer.update(builtins.finalize());
+
+    outer.update(truncated_keccak);
+
+    let bytecodes = contract_definition
+        .program
+        .data
+        .iter()
+        .enumerate()
+        .map(|(i, s)| {
+            StarkHash::from_hex_str(&*s).with_context(|| format!("Invalid bytecode at index {i}"))
+        })
+        .try_fold(HashChain::default(), |mut hc, next| {
+            hc.update(next?);
+            Result::<_, Error>::Ok(hc)
+        })
+        .context("Failed to process contract_definition.program.data")?;
+
+    outer.update(bytecodes.finalize());
+
+    Ok(outer.finalize())
+}
+
+/// HashChain is the structure used over at cairo side to represent the hash construction needed
+/// for computing the contract hash.
+///
+/// Empty hash chained value equals `H(0, 0)` where `H` is the pedersen_hash function, and the
+/// second value is the number of values hashed together in this chain. For other values, the
+/// accumulator is on each update replaced with the `H(hash, value)` and the number of count
+/// incremented by one.
+struct HashChain {
+    hash: StarkHash,
+    count: usize,
+}
```

Comment:
Is this something we can optimise by building it into pedersen? To stay in the faster projection thingy.

Reply:
I tried looking at that, but the affine x coordinate is needed for the bits, and that's widely different for projective vs. affine.